freeCodeCamp builder

mdp opened the pull request:
This is now working and includes:

- A Dockerfile for building the zim file (see Readme)
- A Makefile for fetching, preparing the curriculum, building the Vite app, and packaging it in a zim file
- Support for 9 of the Javascript courses
- Support for all the languages currently supported by FCC
- Basic searching support in the zim file (This is an SPA, so this is handled as redirects)

rgaudin reviewed (changes requested):
Thank you @mdp!

This is a great first step; I think the output ZIM is quite usable ; it's very promising.
As you know, one of the main objective of this iteration is to get this rolling in our Zimfarm.
So below is a list of improvements we'd like to see towards this end: Zimfarm creating FCC ZIMs.

- Refactor so we can pass all parameters to the CLI. ZF works off a single `CMD` which we construct based on options. Makefile currently constains a static list of courses and other important options that require setting environment variables to change.
- ZIM Metadata must properly be set:
  - `Name` must respect our [naming convention](https://github.com/openzim/overview/wiki/Naming-Convention) if set automatically or be user-set.
  - `Language` must be set (as you need the FCC langage string, you'd need to map that to ISO-639-3)
  - `Title` must be set
  - `Description` must be set
  - `Scraper` should be set (`fcc2zim vX`)
- Filename should respect that same convention. Usually `{Name}_{period}.zim`. Period being `%Y-%m`.
- Remove the redirects in **build**. Those are incorrect at the moment: target and source are the same and it confuses libkiwix (I shall open a ticket upstream). Also, it's not usefull because of how you built the client app. Last, entries like `index.html#/english/regular-expressions/extract-matches` cannot be used because anything after the `#` is considered the fragment by the browser and thus not sent to the server (libkiwix).
- Ability to no specify the courses and get all matching the filter

Here's how the cli could look like

```sh
fcc2zim --output /output \
	--name fcc_en_javascript \
	--title "FreeCodeCamp Javascript" \
	--description "FCC Javascript Courses" \
	--lang english \
	--filter 02-javascript \
	--courses "regular-expressions basic-javascript basic-data-structures debugging functional-programming object-oriented-programming basic-algorithm-scripting intermediate-algorithm-scripting javascript-algorithms-and-data-structures-projects"
```

---

In addition to those, I'm listing a few things we'd eventually do to to ease maintenance and align with our standard but that we don't expect you to do.

- Use black code formatting with isort and flake8 with a workflow checking this (similar to https://github.com/openzim/cms/blob/main/.github/workflows/qa.yml)
- Refactor client to use mardown files at runtime instead of as assets. Using them as assets make them a dependency to build the client which force the whole thing to be written to the filesystem first. Instead, we'd add them directly to the ZIM and the code would include them via URL.
- Preprocess the mardown files into HTML before putting them into the ZIM. Depends on previous one. We'd need to check the code would be happy with HTML but preprocessing this in python would save some tiny bits of CPU on the running clients.
- Build the client app directly in a multi-stage Dockerfile as it wouldn't be dependent on content anymore.
- use pyyaml instead of frontmatter in the Python code
- Change fetch step to be solely about fetching source and move the filtering to prebuild
- Fusion prebuild and build to apply the filter, get the appropriate files and directly put them into the ZIM
- Remove path.py dependency (covered by `pathlib`)
- `python-magic` and `libzim` not required as dependencies (scraperlib deps not used directly)
- Use the markdown README in Python package setup so we have a single source that's GH friendly
- Use a python source image so we can pin it properly. If client refactor done, use a stage to build it, otherwise install node via nvx
- Make use of that FCC logo that's bundled in the client. Probably via some header line?

mdp commented:
Thanks for the feedback. Most of this is pretty easy to address but I have a question around the structuring of the client app and content.

Currently we have a Vuejs client that takes in a URL fragment, which is maps to course content and pulls it in the background on a change. With regards to the fragment, can we fix this we query params? Something like `index.html?course=english%2FTregular-expressions%2Fextract-matches`. Would it then be possible to use a redirect? Unfortunatley Vue routing doesn't really provide an option for building out a single page application for each Challenge ("english/regular-expressions/extract-matches/index.html"), which would be pretty redundant, but offers either a hash mode, or HTML5 mode (which requires a catch-all route to the SPA index.html - https://router.vuejs.org/guide/essentials/history-mode.html#html5-mode). I'm pretty sure using a catch-all is not an option in Libzim.

I'll start on the rest of the feedback today.

rgaudin commented:
> I'm pretty sure using a catch-all is not an option in Libzim.

Indeed it wouldn't work.
The query-string instead of the fragment would suffer from the same limitation.

I don't see any way to do it using a ZIM redirect because those are ZIM-entry to ZIM-entry redirect and no request context is sent to the target.

I think the easiest solution, which is not the prettiest but is one we've followed in the past, is to create stub ZIM entries for each course (or curriculum – what did you have in mind?)
Entry would be a minimal HTML that simply declares a redirect (HTTP one this time) – via meta/JS – to the fragment-based URL of the content.

This would allow those entries to be served by the Suggestion feature, as well as the Random one. We could even (in a later optimization), give the text in the corresponding markdown to xapian so that those entries are returned by the full text engine.

mdp commented:
I think this in decent shape for another review when you get the chance @rgaudin 

Fixed/Added:
- Multistage docker build client/openzim
- QA workflow with flake8 black and isort
- Client no longer requires curriculum, build then add curriculum later
- One docker command to build client, fetch curriculum, prebuild and zim.
- Redirects are now HTML meta redirects
- Zim metadata should now be fixed and valid (version is also currently v1.0.0)
- Mapped FCC languages to ISO-639-3 (See outstanding items about zh and zh-tw)
- Removed unneccessary dependencies (eg. Pathlib, libzim) and unnecessary imports
- Pyyaml instead of `frontmatter` for markdown parsing of frontmatter.
- Readme added to setup.py

Outstanding items:
- FCC has `chinese` and `chineses-traditional` which would normally map to zh-cn and zh-tw respectively, but not sure about this with 3 lettter ISO.
- I'm unsure about making use of the FCC logo that's bundled in the client, do you mean in the zim file?

Thanks

rgaudin commented:
> * FCC has `chinese` and `chineses-traditional` which would normally map to zh-cn and zh-tw respectively, but not sure about this with 3 lettter ISO.

I have zero knowledge of the Chinese languages so this may be completely off but from Wikipedia and Etnologue, I'd map `chinese=cmn` and `chinese-traditional=lzh`.

So:

```py
FCCLangMap = {
    "cmn": "chinese",
    "lzh": "chinese-traditional",
    ...
}
```

kelson42 commented:
@mdp I guess it would be better to have a lock file js deps. Otherwise no garanty that we get the same result by running twice the scraper.

mdp commented:
> @mdp I guess it would be better to have a lock file for js deps. Otherwise no garanty that we get the same result by running twice the scraper.

I'm using 'yarn' with the --frozen-lockfile in the latest Dockefile, but let me know if that's not working. Unless something else is happening at another level, the Vue application being built with Vite should be reproducible as well. I verfied it by deleting the dependencies, reinstalling from the lockfile and doing a shasum on the resulting build and it came out the same.

rgaudin commented on client/src/pages/HomePage.vue:
```diff
@@ -1,21 +1,18 @@
-    `../assets/fcc/locales/${language}/intro.json`
-  )
-).default['javascript-algorithms-and-data-structures']
+const locales = (await (await fetch(`fcc/locales/${language}/intro.json`)).json())['javascript-algorithms-and-data-structures']
```
Restricting this to `javascript-algorithms-and-data-structures` here prevents me from including any course that is not part of this. Is this intentional? I can build the ZIM and have the content but it won't be displayed on the homepage…

mdp replied:
This was leftover from the first pass, and fixed in the latest. The new URL scheme matches the structure of FCC's content =>
`/lang/superblock/course/challenge`

rgaudin commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
         for file in fileList:
             print(file)
             path = os.path.relpath(file, source_dir)
             creator.add_item_for(path, fpath=file)
```
All but the home page should be `is_front=False`

mdp replied:
When I do this, I can't search for content by title. Is there another way to handle this?

rgaudin reviewed (changes requested):
Thank you @mdp  ; those are great improvements.

I've added some inline comments ; mostly about python-stuff.
A few general comments:

- README should explain where/how to find the params for `--course`.
- Not enough feedback: when starting the tool, the download of the file and file extraction happens without any feedback.
- User inputs should be checked (to some extent): An incorrect `--language` only crashes in build. Similarly, an incorrect course should fail in fetch as “not found”.

CLI parameters are confusing:

- `--outdir` is not an output folder but the client's one. And actually an untold subfolder of it.
- `--clientdir` very similar (and currently interdependant) with `outdir`.

I think we can save flexibility here and hard code paths to the client.

`--outzim` with the presence of `--outdir` is confusing but if we get rid of the former, we should be OK.

Instructions for building the client are left out (or must be copied from Dockerfile).
As @kelson42, we need the `yarn.lock` file in the repo, otherwise there's no guarantee that, over time, the dependencies pulled are identical.

Finally, there's the issue about restricting items on the home page.

I've also seen cases of blank pages with a `No description in challenge` in the console.

Last, the suggestions don't work but that seems to be due to a bug in kiwix-serve. I'll come back once this is sorted out.

mdp commented on openzim/fcctozim/fetch.py:
```diff
@@ -5,37 +5,31 @@
 
     # Don't redownload the file if we already have it (it's a large file)
     if force or not os.path.exists(zip_path):
         r = requests.get(url, allow_redirects=True)
```
Resolved in 309eec193d38b41a2deb326079ec06a6105fa308

mdp commented on openzim/fcctozim/fetch.py:
```diff
@@ -5,37 +5,31 @@
     if force or not os.path.exists(zip_path):
         r = requests.get(url, allow_redirects=True)
-        with open(zip_path, 'wb') as f:
+        with open(zip_path, "wb") as f:
```
Resolved in 309eec193d38b41a2deb326079ec06a6105fa308

mdp commented on openzim/fcctozim/fetch.py:
```diff
@@ -5,37 +5,31 @@
-                    zip_ref.extract(zip_info, curriculum_path)
-                    file_count = file_count + 1
+                zip_ref.extract(zip_info, curriculum_path)
+                file_count = file_count + 1
```
Resolved in 309eec193d38b41a2deb326079ec06a6105fa308

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -57,36 +71,45 @@ def write_course_to_path(course_list, course_slug: str, outdir: str, language="e
-    for COURSE in course_list.split(','):
+    lang = FCCLangMap[arguments.language]
+    tmpdir = arguments.tmpdir or default_tmp_path
+    curriculum_dir = os.path.join(
```
Sorry, I read this wrong in previous review and removed pathlib in favor of os.path. I've removed all os.* usage in 309eec193d38b41a2deb326079ec06a6105fa308

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -57,36 +71,45 @@ def write_course_to_path(course_list, course_slug: str, outdir: str, language="e
+    locales_dir = os.path.join(
+        tmpdir, f"curriculum/freeCodeCamp-main/client/i18n/locales/{lang}"
+    )
+    print(tmpdir, curriculum_dir)
```
Resolved in 309eec193d38b41a2deb326079ec06a6105fa308

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -57,36 +71,45 @@ def write_course_to_path(course_list, course_slug: str, outdir: str, language="e
+    )
+    print(tmpdir, curriculum_dir)
+
+    for COURSE in course_list.split(","):
```
Sorry, this was a leftover from debugging with a constant. Resolved in 309eec193d38b41a2deb326079ec06a6105fa308

mdp commented on openzim/fcctozim/__init__.py:
```diff
@@ -11,7 +11,21 @@
 TMP_FOLDER = "tmp"
 TMP_FOLDER_PATH = path(TMP_FOLDER)
```
Resolved in https://github.com/openzim/freecodecamp/commit/309eec193d38b41a2deb326079ec06a6105fa308#diff-87f9f8b7ea3c4e3b884c25099af1ee8de051b36eada9f56d4111e4964d427503

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -57,36 +71,45 @@ def write_course_to_path(course_list, course_slug: str, outdir: str, language="e
 
-        course_list: List[id: str, path: str, title: str] = []
+        course_list: List[str, str, str] = []
+        # List[id, path, title]
```
Resolved by e00860503

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -3,51 +3,65 @@
+default_tmp_path = "./tmp"
+
+
+def readFrontmatter(filename: str):
```
Resolved in 3f5da518

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -3,51 +3,65 @@
+def write_course_to_path(
+    course_list, course_slug: str, outdir: str, language="english"
+):
+    course_dir = f"{outdir}/curriculum/{language}/{course_slug}"
```
Now all pathlib, resolved in 309eec193d38b41a2deb326079ec06a6105fa308

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -3,51 +3,65 @@
 def write_locales_to_path(source_dir: str, outdir: str, language="english"):
-    locales_dir = f'{outdir}/locales/{language}'
+    locales_dir = f"{outdir}/locales/{language}"
     print(source_dir, locales_dir)
```
Removed in the latest

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -3,51 +3,65 @@
+import yaml
+from fcctozim import FCCLangMap
+
+default_tmp_path = "./tmp"
```
Resolved in 309eec193

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -57,36 +71,45 @@ def write_course_to_path(course_list, course_slug: str, outdir: str, language="e
-            id = info['attributes']['id']
-            title = info['attributes']['title']
+            info = readFrontmatter(file)
+            id = info["id"]
```
Resolved in 29100645

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -57,36 +71,45 @@ def write_course_to_path(course_list, course_slug: str, outdir: str, language="e
+            id = info["id"]
+            title = info["title"]
             try:
                 if ids.index(id) > -1:
```
Resolved in 29100645

mdp commented on openzim/fcctozim/prebuild.py:
```diff
@@ -3,51 +3,65 @@
     index = json.load(f)
-    if not index.get(language): index[language] = []
-    if not slug in index[language]:
+    if not index.get(language):
```
Resolved in 29100645

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
-    rootPath = os.path.join(source_dir, 'index.html')
+
+def build(arguments):
+    clientdir = arguments.clientdir
```
All Pathlib in 309eec193

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
+
+    # Make sure the outpath directory exists
+    outpathExists = os.path.exists(os.path.dirname(outpath))
+    if not outpathExists:
```
Refatored with https://github.com/openzim/freecodecamp/commit/309eec193d38b41a2deb326079ec06a6105fa308#diff-b914a73f366994735aa1a6acad1ce82ec8faca47a68294c836f987674ffa0f90R59

`pathlib.Path(outpath).mkdir(parents=True, exist_ok=True)` to avoid the if check

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
     ) as creator:
         for file in fileList:
             print(file)
             path = os.path.relpath(file, source_dir)
```
Resolved with https://github.com/openzim/freecodecamp/commit/309eec193d38b41a2deb326079ec06a6105fa308#diff-b914a73f366994735aa1a6acad1ce82ec8faca47a68294c836f987674ffa0f90R72

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
 def build_curriculum_redirects(clientdir, language):
-    index_json_path = os.path.join(clientdir, 'src/assets/fcc/index.json')
+    fcc_lang = FCCLangMap[language]
+    index_json_path = os.path.join(clientdir, "fcc/index.json")
```
Resolved in https://github.com/openzim/freecodecamp/commit/309eec193d38b41a2deb326079ec06a6105fa308#diff-b914a73f366994735aa1a6acad1ce82ec8faca47a68294c836f987674ffa0f90R13

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
+        f"Building {clientdir} for {language} => {outpath} - Version: {VERSION}"
+    )
+
+    source_dir = os.path.join(clientdir)
```
Removed and cleaned up for clarity in [2f53afdd0dd303cdaed1e40f323ee8992d06a211](https://github.com/openzim/freecodecamp/pull/5/commits/2f53afdd0dd303cdaed1e40f323ee8992d06a211)

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
+        )
         with open(meta_json_path) as meta_json_str:
-            challenges = json.load(meta_json_str)['challenges']
+            challenges = json.load(meta_json_str)["challenges"]
```
Resolved in https://github.com/openzim/freecodecamp/pull/5/commits/2f53afdd0dd303cdaed1e40f323ee8992d06a211

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
             creator.add_item_for(path, fpath=file)
-        
 
         for course_page in build_curriculum_redirects(clientdir, language):
```
Resolved in https://github.com/openzim/freecodecamp/pull/5/commits/2f53afdd0dd303cdaed1e40f323ee8992d06a211

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
+            redirect_url = (
+                len(redirect_path.split("/")) - 1
+            ) * "../" + f"index.html#{course_page[0]}"
+            content = f'<html><head><title>{title}</title><meta http-equiv="refresh" content="0;URL=\'{redirect_url}\'" /></head><body></body></html>'  # noqa: E501
```
Resolved in https://github.com/openzim/freecodecamp/pull/5/commits/2f53afdd0dd303cdaed1e40f323ee8992d06a211

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
+            redirect_url = (
+                len(redirect_path.split("/")) - 1
+            ) * "../" + f"index.html#{course_page[0]}"
+            content = f'<html><head><title>{title}</title><meta http-equiv="refresh" content="0;URL=\'{redirect_url}\'" /></head><body></body></html>'  # noqa: E501
```
Resolved in https://github.com/openzim/freecodecamp/pull/5/commits/2f53afdd0dd303cdaed1e40f323ee8992d06a211

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,66 +1,95 @@
-        language=language
-    )
\ No newline at end of file
+            redirect_path = f"redirect/{course_page[0]}"
```
Not reaquired, cleaned up and removed in 26d269c8

mdp commented:
@rgaudin I think there's just the remainin question about is_front and being able to search for Challenges, but all the other items should be resolved. Thanks for all your help on this. It's been several years since I've done Python (and then very little), sorry for the less than Pythonic code.

rgaudin reviewed (changes requested):
Thank you @mdp ! I figured when I saw Path.py! Brought me back a few years back 😉

Most of the inline comments are suggestions towards _our convention_. Feel free to discard them.

I'd like us to check the CSS hack though. I think checking that the client was built before creating the ZIM would also help.

Regarding `is_front` ; I've explained inline. As there is a bug in libkiwix, we have to choose between suggestions and random ATM but we'll fix that.

Once this is merged, I'd like you to provide a list of what the scraper can and can't do (curriculum/courses). As suggested earlier, we need to know what options to pass `--course`.

I've noticed that most (if not all) of `python-for-everybody` fail to load with a “No description in challenge” exception (already mentioned). Their frontmatter doesn't include a `description` so that's the reason but they do have content and skipping the exception just reveals broken challenges.

If we are to create FCC ZIM files, we need to know the limits of this tool and at the moment it's not defined.

rgaudin commented:
Just saw another glitch ; in `python-for-everybody`, a number of challenges like `networking-text-processing` have titles that include a colon `:` which breaks the YAML parsing.

See `/fcc/curriculum/english/scientific-computing-with-python/python-for-everybody/networking-text-processing.md`:

```yaml
---
id: 5e7b9f0c0b6c005b0e76f074
title: 'Networking: Text Processing'
challengeType: 11
videoId: Pv_pJgVu8WI
bilibiliIds:
  aid: 804442498
  bvid: BV16y4y1j7WW
  cid: 377329124
dashedName: networking-text-processing
---
```

The rendered title is `'Networking`

<img width="568" alt="Screenshot 2023-07-04 at 09 44 45" src="https://github.com/openzim/freecodecamp/assets/57929/7303f7c5-ce78-4668-b7ee-e23b2829ae9b">

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,95 +1,98 @@
         Title=title,
+        Publisher="Kiwix",
+        Date=datetime.now(),
+        Creator=creator,
```
Resolved in 6317096c4c7aae22ac89316bfe87dff2c67a7271

mdp commented on openzim/fcctozim/build.py:
```diff
@@ -1,95 +1,98 @@
+    with Creator(outpath, main_path.as_posix()).config_metadata(
         Name=name,
         Title=title,
+        Publisher="Kiwix",
```
Resolved in 6317096c4c7aae22ac89316bfe87dff2c67a7271

mdp commented on client/src/style.css:
```diff
@@ -22,6 +22,8 @@ body {
   -moz-osx-font-smoothing: grayscale;
   -webkit-text-size-adjust: 100%;
+  /* Kiwix iFrame will overlay 5em on the bottom */
+  margin-bottom: 7em;
```
This is scrolled all the way to the bottom, but the bottom bar overlaps the content.
![image](https://github.com/openzim/freecodecamp/assets/2868/62238957-9cc6-4e04-b4ec-d7662a2c06b1)

It's not a huge issue, but I can't seem to easily fix it on my end. This is in Chrome on Linux (Version 114.0.5735.198 (Official Build) (64-bit))

benoit74 replied:
I saw the same issue with `kolibri`.
Looks like this is only a `kiwix-js` issue from my experience (no issue on kiwixserve typically).
@Jaifroid : are you aware of anything like this?

rgaudin replied:
OK then we can drop that hack. @benoit74 please open a kiwix-js ticket

Jaifroid replied:
Sorry to be slow to reply -- didn't see the thread till now. We do have an issue on Kiwix JS with the bottom bar hiding content. No need to create a new one: https://github.com/kiwix/kiwix-js/issues/435 and https://github.com/kiwix/kiwix-js/issues/337.

I'd like to test a ZIM on Kiwix JS PWA (https://pwa.kiwix.org), which has some hackish workaround code with the bottom bar issue.

benoit74 replied:
@Jaifroid : from my tests with very few ZIMs on Kiwix JS PWA, I confirm the issue is "solved" on this platform, the bottom bar fades out as we slide down. I don't know what would happen if there is no slider because the page perfectly fits the viewport (extreme corner case, but, you know ...)

benoit74 commented:
I would like to thank you @mdp, this is really great work I think. The last mile seems almost here now.

I appreciate especially the fact that this scrapper proves it is totally feasible / easy to host a Vite (Vue.JS here, but probably does not matter much indeed) application inside a Zim. We briefly discussed about it with @kelson42 and @rgaudin few days ago for the [UI enhancements of kolibri](https://github.com/openzim/kolibri/issues/42), for which I was tempted to migrate to Vue.JS for simpler/smaller UI code once the UI becomes a bit dynamic/complex. I believe that you gave a good answer to my interrogations ^^

rgaudin reviewed (commented):
Those changes are all good. Looking at the unresolved discussions now

mdp commented:
Sorry, just now getting back to this. Tackling the current build break, which looks like a minor directory issue.

mdp commented:
@rgaudin With regards to the [colon issue](https://github.com/openzim/freecodecamp/pull/5#issuecomment-1619706243) I need to play around with Python's YAML parsing and figure out if there's an easy fix on this one. I'll try to have this one fixed today as well as the tests.

mdp commented:
I just pushed a build fix. Merge sounds good. I'll start on the colon issue now.
